File: frontend/src/api/task.ts
import { apiFetch } from "./client";

// Інтерфейс задачі, що повністю відповідає схемі з нашого FastAPI бекенду
export interface Task {
  id: number;
  project_id: number;
  status_id: number;
  title: string;
  description: string | null;
  priority: number; // 1-Low, 2-Medium, 3-High
  created_at: string;
  updated_at: string;
}

/**
 * Отримання списку всіх задач для конкретного проекту (дошки)
 */
export async function fetchTasksByProject(projectId: number): Promise<Task[]> {
  return apiFetch<Task[]>(`tasks/project/${projectId}`);
}

/**
 * Створення нової картки задачі всередині конкретної дошки та колонки
 */
export async function createTask(taskData: {
  project_id: number;
  status_id: number;
  title: string;
  description?: string;
  priority?: number;
}): Promise<Task> {
  return apiFetch<Task>("tasks/", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(taskData),
  });
}

export async function updateTask(
  taskId: number,
  taskData: { status_id?: number; priority?: number },
): Promise<Task> {
  return apiFetch<Task>(`tasks/${taskId}`, {
    method: "PATCH",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(taskData),
  });
}

export async function fetchPriorityLabels(): Promise<Record<number, string>> {
  return apiFetch<Record<number, string>>("tasks/priorities");
}

File: backend/app/api/routers/task.py
from typing import Optional
from fastapi import APIRouter, Depends, HTTPException, status
from pydantic import BaseModel
from sqlmodel import Session

from app.core.db import get_session
from app.services import task as task_service

router = APIRouter(prefix="/tasks", tags=["Tasks"])


class TaskCreate(BaseModel):
    project_id: int
    status_id: int
    title: str
    description: Optional[str] = None
    priority: int = 2


class TaskUpdate(BaseModel):
    status_id: Optional[int] = None
    priority: Optional[int] = None


@router.post("/", status_code=status.HTTP_201_CREATED)
def create_new_task(task_data: TaskCreate, session: Session = Depends(get_session)):
    """Створення нової картки задачі"""
    return task_service.create_task(
        session=session,
        project_id=task_data.project_id,
        status_id=task_data.status_id,
        title=task_data.title,
        description=task_data.description,
        priority=task_data.priority,
    )


@router.get("/project/{project_id}")
def read_project_tasks(
    project_id: int, status_id: Optional[int] = None, session: Session = Depends(get_session)
):
    """Отримання всіх задач проекту з автоматичним сортуванням (Kanban-дошка)"""
    return task_service.get_tasks_by_project(
        session=session, project_id=project_id, status_id=status_id
    )


@router.patch("/{task_id}")
def update_task(
    task_id: int, update_data: TaskUpdate, session: Session = Depends(get_session)
):
    """Оновлення колонки або пріоритету (наприклад, при Drag & Drop перетягуванні)"""
    updated_task = task_service.update_task_status_or_priority(
        session=session,
        task_id=task_id,
        status_id=update_data.status_id,
        priority=update_data.priority,
    )
    if not updated_task:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Задачу не знайдено")
    return updated_task
